### ISSUE TITLE
Freeze of GlobalSimilarity.py

### ISSUE THREAD

**Issue by bartolsthoorn** [cosmo-epfl/glosim2#1]
Hi, when I run `GlobalSimilarity.py` it seems to run but it just freezes and nothing happens, this is the output:
```
root@0f7ecc515774:~/glosim2# python GlobalSimilarity.py ../100.xyz --nocenters 1
 WARNING: libAtoms::Hello World: environment variable OMP_STACKSIZE not set explicitly. The default value - system and compiler dependent - may be too small for some applications.
_____/\\\\\\\\\\\\__/\\\\\\_____________________________________________________________/\\\\\\\\\_____        
 ___/\\\//////////__\////\\\___________________________________________________________/\\\///////\\\___       
  __/\\\________________\/\\\________________________________/\\\______________________\///______\//\\\__      
   _\/\\\____/\\\\\\\____\/\\\________/\\\\\_____/\\\\\\\\\\_\///_____/\\\\\__/\\\\\______________/\\\/___     
    _\/\\\___\/////\\\____\/\\\______/\\\///\\\__\/\\\//////___/\\\__/\\\///\\\\\///\\\_________/\\\//_____    
     _\/\\\_______\/\\\____\/\\\_____/\\\__\//\\\_\/\\\\\\\\\\_\/\\\_\/\\\_\//\\\__\/\\\______/\\\//________   
      _\/\\\_______\/\\\____\/\\\____\//\\\__/\\\__\////////\\\_\/\\\_\/\\\__\/\\\__\/\\\____/\\\/___________  
       _\//\\\\\\\\\\\\/___/\\\\\\\\\__\///\\\\\/____/\\\\\\\\\\_\/\\\_\/\\\__\/\\\__\/\\\___/\\\\\\\\\\\\\\\_ 
        __\////////////____\/////////_____\/////_____\//////////__\///__\///___\///___\///___\///////////////__
Sat Mar 17 08:33:35 2018
Start Computing the global average kernel of ../100.xyz
using output prefix = ../100-n8-l6-c3.5-g0.5-cw1.0-cotw0.5-average-zeta2-nonorm
Reading 100 input atomic structure from ../100.xyz: done 00:00:00 (h:m:s)
Compute soap and environmental kernels with a pool of 4 workers and 4 threads over 10 chunks: 00:00:00 (h:m:s)
listener  False
Env kernels:   0%|                                                                                                                                             | 0/5050 [00:00<?, ?it/s]
```

This is in the quippy docker and with `100.xyz` contains 100 libatoms xyz crystal structures.

Do you maybe know what is going on?

**Comment by bartolsthoorn** [comment]
Sometimes I get this error:
```
Exception AttributeError: "'tqdm' object has no attribute 'pos'" in <object repr() failed> ignored
Traceback (most recent call last):
  File "GlobalSimilarity.py", line 249, in <module>
    **soap_params)
  File "GlobalSimilarity.py", line 114, in get_environmentalKernels
    environmentalKernels = get_environmentalKernels_mt_mp_chunks(**kargs)
  File "/root/glosim2/libmatch/environmental_kernel.py", line 658, in get_environmentalKernels_mt_mp_chunks
    results = pool.run()
  File "/root/glosim2/libmatch/environmental_kernel.py", line 526, in run
    res = self.pool.map(self.func, self.chunks)
  File "/usr/local/lib/python2.7/multiprocessing/pool.py", line 253, in map
    return self.map_async(func, iterable, chunksize).get()
  File "/usr/local/lib/python2.7/multiprocessing/pool.py", line 572, in get
    raise self._value
AttributeError: 'tqdm' object has no attribute 'pos'
```
It might be related to https://github.com/tqdm/tqdm/issues/323 (which is still an open issue)

**Comment by bartolsthoorn** [comment]
With 1 process/chunk/thread
```
get_environmentalKernels(islow_memory=False, nthreads=1, nprocess=1, nchunks=1, **soap_params)
```
and in jupyter notebook it works. So I'll close this for now.

**Comment by felixmusil** [comment]
Hi, I just tried to run on my set up and it seems the problems comes from the threads and processes at the same time. So for the moment you can try to run with one of them set to 1 (threads I would recommend). 
Thanks for reporting the bug.

**Comment by bartolsthoorn** [comment]
Thank you for your reply.

Does this code support periodic structures (crystal structures) by the way?

Also, this is a minimal example in Jupyter notebook that I use to test performance:
```python
%%time
import quippy
from GlobalSimilarity import get_globalKernel, get_environmentalKernels

from libmatch.chemical_kernel import Atoms2ChemicalKernelmat, deltaKernel

atoms = quippy.AtomsList('../4000.xyz', start=0, stop=10)

soap_params = {
    'atoms': atoms, 'centerweight': 1.0, 'gaussian_width': 0.5,
    'cutoff': 3.5, 'cutoff_transition_width': 0.5,
    'nmax': 8, 'lmax': 6, 'chemicalKernel': deltaKernel,
    'chemicalProjection': None,
    'chem_channels': True, 'nocenters': [1],'is_fast_average': False,
}

environmentalKernels = get_environmentalKernels(islow_memory=False, nthreads=1, nprocess=1, nchunks=1, **soap_params)
```

With just 10 crystal structures, I get:
`nthreads=1, nprocess=1, nchunks=1 # => Wall time: 39.9 s`
`nthreads=2, nprocess=1, nchunks=1  #=> Wall time: 39.7 s`
`nthreads=1, nprocess=1, nchunks=2  #=> Freeze at 73%`
`nthreads=1, nprocess=2, nchunks=1  #=> Freeze at 0%`

To summarize, the only thing that works is increasing `nthreads`, but it seems this does not actually speed up the calculations.

Another thing I noticed is that if I take only take 10 materials it takes 4 seconds per crystal structure. If I call `get_environmentalKernels` with 4000 materials it takes about 40 seconds per second crystal. I am not sure why it is slower, because `get_environmentalKernels` should just calculate the kernels for each material independently?

The crystal structures I am using have about 100 atoms in the unit cell on average. To process 4000 structures it would take about 46 hours to complete `get_environmentalKernels` at the current speed..

**Comment by felixmusil** [comment]
Thanks for your comments. Bear in mind that this implementation is a first attempt to make the whole procedure faster (compared to glosim) and I am well aware of its current limitations. I have a clear idea on how to improve it but I lack the time at the moment to reimplement all the possible settings.

I am surprised by your timings and that it takes so much time since I just computed the similarity between 7000 structures with about 50 atoms per cell in ~3hours. Maybe the docker container is not the best evironment to run heavy duty work. 

Yes periodic boundary conditions are applied by default, so if you don't want them you should explicitly state it in the .xyz file as expected by the extended xyz format.

Let me summarize the action of these parameters to guide you:
* Threads should help with large frames by splitting the computation of individual environmental matrices.
* Process trivially splits the computation of the soap vectors and if you have chunks also for the environmental similarity matrix. It should be very effective for the computation of the soap (low communications) while it is not so clear on the chunks (large overhead).
* Chunks divides the computation of the similarity matrix in parts and each part can be handled by a different process. It is mainly intended for when the environmental similarity matrix does not fit in memory.

**Comment by bartolsthoorn** [comment]
Thank you for your reply, that summary really helpful! It sounds like increasing `nprocess` would be most effective.

I think Docker should not be the reason for significant overhead. More importantly, with 1 process I am only using 1 core but I have 64 available, so if the soap vectors calculations are split trivially that should be able to cut from 46 hours to 43 minutes, that would be great! :smile: 

The problem is that if I set for example `nthreads=1, nprocess=8, nchunks=1`, it just stalls at 0%:
![screenshot from 2018-03-18 08-43-13](https://user-images.githubusercontent.com/106586/37563765-6e210876-2a88-11e8-95e1-03d8ebcd8dd9.png)
(traceback: https://gist.github.com/bartolsthoorn/07e58ce13253e17fdc332a1b25c21261)

I will try to see if I can debug why that is happening, because in the end I would like to scale up to 25,000 structures. Finally, the goal is to extract a train and test kernel from the complete matrix and do kernel ridge regression, roughly following the steps below (will expand into train/validation/test):
```python
a = k + l * np.eye(len(Y))
coefficients = scipy.linalg.solve(a,Y, sym_pos=True,overwrite_a=True)
predictions = np.dot(k,coefficients)
print('MAE', np.mean(np.abs(Y - predictions)))
```

Another short question: In which scenario would setting `nocenters=[1]` make sense? I assume that it speeds up the calculations by skipping all the hydrogen and its positions should not be significant factors for most properties that people predict?

**Comment by bartolsthoorn** [comment]
I found a work-around for the bug, by setting the following before running `get_Soaps` with `nprocess` greater than 1:
```python
from tqdm import tqdm
tqdm.monitor_interval = 0
```
Source: https://github.com/tqdm/tqdm/issues/481

Now I can increase `nprocess` without stalling and it looks like multiple cores are used.

**Comment by bartolsthoorn** [comment]
I have narrowed down the problem a bit: the program hangs if you use `libmatch.soap.get_soap` within a `multiprocessing.Pool`. If you do the same outside a Pool it just works, really weird.

**Comment by bartolsthoorn** [comment]
Now that the libatoms website is online again I installed QUIPPY without docker and the issue of only being able to use one CPU is gone!

So this issue is now resolved.
